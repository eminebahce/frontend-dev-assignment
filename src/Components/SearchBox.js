import React, {Component} from 'react';
import PropTypes from 'prop-types';
import axios from 'axios';
import _ from 'lodash';
import Input from './Input';
import DropdownList from './DropdownList';
import styled, { createGlobalStyle } from 'styled-components'
import Suggestions from './Suggestions';

class SearchBox extends Component {
        state = {
            active: false,
            search: '',
            results: []
        };

    setActive = () => {
        this.setState({active: !this.state.active})
    }

    handleChange = (event) => {
        this.setState({search: event.target.value})
    }

    resetSearch = () => {
        this.setState({search: ''})
    }

    delayedCallBack = _.debounce(()=>{
        axios.get(`http://localhost:5000/search?q=${this.state.search}`)
            .then(response =>
                this.setState({results: response.data.suggestions.filter(searchedItem => searchedItem.searchterm.includes(this.state.search))}))
            .catch(error => console.log(error))
    }, 1000);


    render() {
        return (
                    <div data-test="search-box">
                        <form>
                        <Input className="input-field"
                               data-test="input-search"
                               type="text"
                               name="input"
                               aria-label="input search"
                               value={this.state.search}
                               onKeyUp={this.delayedCallBack}
                               onFocus={this.setActive}
                               onChange={(e) => this.handleChange(e)}
                               placeholder={this.props.placeholder}
                        />
                        {this.state.search &&
                        <button className="clear-input"
                                type="reset"
                                onClick={this.resetSearch} >
                            Clear
                        </button>}
                        <DropdownList>
                            <Suggestions suggestionsList={this.state.results}/>
                        </DropdownList>
                        </form>
                    </div>
        )
    }
}

SearchBox.propTypes = {
    setActive: PropTypes.func,
    handleChange: PropTypes.func,
    resetSearch: PropTypes.func,
    delayedCallBack: PropTypes.func
};

export default SearchBox;
