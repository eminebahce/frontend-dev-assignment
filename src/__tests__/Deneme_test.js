import React from 'react';
import {shallow} from 'enzyme';
import checkPropTypes from 'check-prop-types';
import SearchBox from '../Components/SearchBox';
import jest from 'jest-mock';
import renderer from 'react-test-renderer';

describe('<SearchBox /> component', () => {
    beforeEach(() => {
        const props = {
            setActive: () => {},
            handleChange: () => {},
            resetSearch: () => {},
            delayedCallBack: () => {}
        };
    });

    const checkProps = (component, expectedProps) => {
        const propsErr = checkPropTypes(component.propTypes, expectedProps, 'props', component.name);
        return propsErr;
    };
    it('should render without error', () => {
        const component = shallow(<SearchBox/>);
        const wrapper = component.find(`[data-test='search-box']`);
        expect(wrapper.length).toBe(1)
    });
    it('should render without error', () => {
        const component = shallow(<SearchBox/>);
        const wrapper = component.find(`[data-test='input-search']`);
        expect(wrapper.length).toBe(1)
    });

    describe('Checking propTypes', () => {
        it('should not throw a warning', () => {
            const expectedProps = {
                setActive: () => {},
                handleChange: () => {},
                resetSearch: () => {},
                delayedCallBack: () => {
                }
            };
            const propsError = checkProps(SearchBox, expectedProps);
            expect(propsError).toBeUndefined();
        });
    });

    describe('renders ReviewStarRating element', () => {
        it('should not return error', () => {
            const component = renderer.create(
                <SearchBox/>
            );
            let tree = component.toJSON();
            expect(tree).toMatchSnapshot();
        });
    });

    describe('testing api get fetching', () => {
        const mockAxios = {
            get: jest.fn(() => Promise.resolve({data: {}}))
        }
        mockAxios.create = jest.fn(() => mockAxios);

        const mockData = {
            "search": "default",
            "suggestions": [
                {
                    "searchterm": "heren truien",
                    "nrResults": 1100
                },
                {
                    "searchterm": "dames truien",
                    "nrResults": 1501
                }
            ]
        };
         mockAxios.get.mockImplementationOnce(() => {
            return Promise.resolve({data: mockData})
        });
         expect(mockAxios.get()).toBeTruthy();
    });
});