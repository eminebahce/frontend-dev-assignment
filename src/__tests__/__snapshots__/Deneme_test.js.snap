// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBox /> component renders ReviewStarRating element should not return error 1`] = `
<div
  data-test="search-box"
>
  <form>
    <input
      aria-label="input search"
      className="input-field sc-bdVaJa hbuiSn"
      data-test="input-search"
      name="input"
      onChange={[Function]}
      onFocus={[Function]}
      onKeyUp={[Function]}
      placeholder={undefined}
      type="text"
      value=""
    />
     
    
    <div
      className="sc-bwzfXH jEqMiu"
    >
      <ul
        data-test="suggestions"
      />
    </div>
  </form>
</div>
`;
